core/src/Kokkos_View.hpp

```diff
@@ -565,14 +565,99 @@ class View : public Impl::BasicViewFromTraits<DataType, Properties...>::type {
   }
 
+#ifdef KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK
+ private:
+  template <class Arg>
+  KOKKOS_INLINE_FUNCTION static bool is_in_bounds(size_t extent,
+                                                  const Arg arg) {
+    return static_cast<std::size_t>(arg) < extent;
+  }
+
+  template <class T>
+  KOKKOS_INLINE_FUNCTION static bool is_in_bounds(size_t extent,
+                                                  const std::pair<T, T> arg) {
+    return static_cast<std::size_t>(arg.second) <= extent && arg.first >= 0;
+  }
+
+  template <class T>
+  static bool is_in_bounds(size_t extent, const Kokkos::pair<T, T> arg) {
+    return static_cast<std::size_t>(arg.second) <= extent && arg.first >= 0 &&
+           arg.first <= arg.second;
+  }
+
+  KOKKOS_INLINE_FUNCTION static bool is_in_bounds(size_t /*extent*/,
+                                                  const ALL_t) {
+    return true;
+  }
+
+  template <class RT, class... RP, size_t... Idx, class... Args>
+  KOKKOS_INLINE_FUNCTION static bool subview_extents_valid(
+      const View<RT, RP...>& src_view, std::index_sequence<Idx...>,
+      Args... args) {
+    return (is_in_bounds(src_view.extent(Idx), args) && ... && true);
+  }
+
+  template <class Arg>
+  static void append_error_message(std::stringstream& ss, size_t extent,
+                                   const Arg arg) {
+    ss << arg << " < " << extent;
+  }
+
+  template <class T>
+  static void append_error_message(std::stringstream& ss, size_t extent,
+                                   const std::pair<T, T> arg) {
+    ss << arg.first << " <= " << arg.second << " <= " << extent;
+  }
+
+  template <class T>
+  static void append_error_message(std::stringstream& ss, size_t extent,
+                                   const Kokkos::pair<T, T> arg) {
+    ss << arg.first << " <= " << arg.second << " <= " << extent;
+  }
+
+  static void append_error_message(std::stringstream& ss, size_t /*extent*/,
+                                   const ALL_t) {
+    ss << "Kokkos::ALL";
+  }
+
+  template <class RT, class... RP, size_t... Idx, class Arg0, class... Args>
+  static std::stringstream generate_error_message(
+      const View<RT, RP...>& src_view, std::index_sequence<Idx...>, Arg0 arg0,
+      Args... args) {
+    std::stringstream ss;
+    ss << "Kokkos::subview bounds error (";
+    append_error_message(ss, src_view.extent(0), arg0);
+    ((ss << ", ", append_error_message(ss, src_view.extent(Idx), args)), ...);
+    ss << ')';
+    return ss;
+  }
+#endif
+
+ public:
   //----------------------------------------
   // Compatible subview constructor
   // may assign unmanaged from managed.
 
   template <class RT, class... RP, class Arg0, class... Args>
   KOKKOS_INLINE_FUNCTION View(const View<RT, RP...>& src_view, const Arg0 arg0,
                               Args... args)
-      : base_t(Impl::subview_ctor_tag, src_view, arg0, args...) {}
+      : base_t(Impl::subview_ctor_tag, src_view, arg0, args...) {
```

Comment:
I think we should move this mechanism into BasicView -- we already do index bounds checking there so would make sense to me

Reply:
Moved the check to `BasicView`.